x-pack/plugins/fleet/public/applications/fleet/sections/agents/agent_list_page/components/agent_upgrade_status.tsx

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useMemo } from 'react';
+// import styled from 'styled-components';
+import { FormattedMessage } from '@kbn/i18n-react';
+import { EuiBadge, EuiFlexGroup, EuiFlexItem, EuiIconTip } from '@elastic/eui';
+
+import type { AgentUpgradeDetails } from '../../../../../../../common/types';
+
+function getUpgradeStartDelay(scheduledAt?: string) {
+  return Math.round((Date.parse(scheduledAt || '') - Date.now()) / 36e5);
+}
+
+function getStatusComponents(agentUpgradeDetails?: AgentUpgradeDetails) {
+  switch (agentUpgradeDetails?.state) {
+    case 'UPG_REQUESTED':
+      return {
+        Badge: (
+          <EuiBadge color="accent">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeRequested"
+              defaultMessage="Upgrade requested"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeRequested"
+            defaultMessage="The agent has requested an upgrade."
+          />
+        ),
+      };
+    case 'UPG_SCHEDULED':
+      return {
+        Badge: (
+          <EuiBadge color="accent">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeScheduled"
+              defaultMessage="Upgrade scheduled"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeScheduled"
+            defaultMessage="The agent has been instructed to upgrade. The upgrade will start in {scheduledAt} hours."
+            values={{
+              scheduledAt: getUpgradeStartDelay(agentUpgradeDetails.metadata.scheduled_at),
+            }}
+          />
+        ),
+      };
+    case 'UPG_DOWNLOADING':
+      return {
+        Badge: (
+          <EuiBadge color="accent">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeDownloading"
+              defaultMessage="Upgrade downloading"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeDownloading"
+            defaultMessage="Downloading the new agent artifact version ({downloadPercent}%)."
+            values={{
+              downloadPercent: agentUpgradeDetails?.metadata.download_percent,
```

Comment:
Is `download_percent` going to be always populated? If not, maybe we should set 0 if not set.

Reply:
It should exist and be populated if the state is `UPG_DOWNLOADING`, otherwise I don't think it's set.

Reply:
In saying that, I think it's a really good idea to fall back to 0 in case for some reason the agent doesn't have a download percent. Thanks for the suggesting, will add!

Reply:
May I suggest that if the `download_percent` field is not set OR if it's set to 0, we don't include it in the message? In other words, would it be possible to start including the download percent value in the message only when `download_percent > 0`?  Otherwise we could end up showing a download percent of 0 to the user for the entire time we're in the `UPG_DOWNLOADING` state and then suddenly move to the next state, which might be a bit confusing.